File: exercism/python/tournament/tournament.py
# https://exercism.io/my/solutions/b3b91870e23c4ca7ae4ea6661ece1096


WIN = 'win'
DRAW = 'draw'
LOSS = 'loss'
RESULTS = {WIN, DRAW, LOSS}
POINTS = {WIN: 3, DRAW: 1, LOSS: 0}


class Team():
    def __init__(self, name):
        self.name = name
        self.wins = 0
        self.losses = 0
        self.draws = 0

    def add_result(self, result):
        if result not in RESULTS:
            raise ValueError(f'Result must be one of {RESULTS}')
        if result == WIN:
            self.wins += 1
        elif result == LOSS:
            self.losses += 1
        else:
            self.draws += 1

    @property
    def matches_played(self):
        return self.wins + self.losses + self.draws

    @property
    def points(self):
        return self.wins * POINTS[WIN] + self.losses * POINTS[LOSS] + self.draws * POINTS[DRAW]

    def __lt__(self, other):
        if self.points == other.points:
            return self.name > other.name
        else:
            return self.points < other.points


def tally(rows):

    def read_results(rows):
        other_team_result = {WIN: LOSS, DRAW: DRAW, LOSS: WIN}
        teams = dict()
        for row in rows:
            name1, name2, result1 = row.split(';')
            result2 = other_team_result[result1]
            team1 = teams.setdefault(name1, Team(name1))
            team2 = teams.setdefault(name2, Team(name2))
            team1.add_result(result1)
            team2.add_result(result2)
        return teams

    def table(teams):
        table = ['Team                           | MP |  W |  D |  L |  P']
        for team in sorted(teams.values(), reverse=True):
            row = (
                f'{team.name:30} | '
                f'{team.matches_played:2} | '
                f'{team.wins:2} | '
                f'{team.draws:2} | '
                f'{team.losses:2} | '
                f'{team.points:2}'
            )
            table.append(row)
        return table

    return table(read_results(rows))

File: exercism/go/tournament/tournament.go
// https://exercism.io/my/solutions/9a68f3c1492f4922b0f5090f08966933

package tournament

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"sort"
	"strings"
)

// team stores a team's matches' results
type team struct {
	name    string
	matches int
	draws   int
	losses  int
	wins    int
	points  int
}

// readResults reads, validates and process the input data
func readResults(r io.Reader) (map[string]team, error) {
	teams := map[string]team{}
	s := bufio.NewScanner(r)

	for s.Scan() {
		line := s.Text()
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		matchInfo := strings.Split(line, ";")
		if len(matchInfo) != 3 {
			return teams, fmt.Errorf("invalid format %q (expected: 'team1;team2;result')", line)
		}
		nameTeam1, nameTeam2, result := matchInfo[0], matchInfo[1], matchInfo[2]
		t1, t2 := teams[nameTeam1], teams[nameTeam2]

		switch result {
		case "draw":
			t1.draws++
			t2.draws++
			t1.points++
			t2.points++
		case "loss":
			t1.losses++
			t2.wins++
			t2.points += 3
		case "win":
			t1.wins++
			t2.losses++
			t1.points += 3
		default:
			return teams, errors.New("invalid input")
		}
		t1.matches++
		t2.matches++
		t1.name = nameTeam1
		t2.name = nameTeam2
		teams[nameTeam1], teams[nameTeam2] = t1, t2
	}

	return teams, nil
}

// sortTeams provides a slice of teams, sorted by points and name
func sortTeams(teams map[string]team) []team {
	s := make([]team, 0, len(teams))
	for _, t := range teams {
		s = append(s, t)
	}
	sort.Slice(s, func(i, j int) bool {
		if s[i].points == s[j].points {
			return s[i].name < s[j].name
		}
		return s[i].points > s[j].points
	})
	return s
}

// Tally reads a series of matches results and pretty prints them
func Tally(reader io.Reader, writer io.Writer) error {
	teams, err := readResults(reader)
	if err != nil {
		return err
	}

	header := "Team                           | MP |  W |  D |  L |  P\n"
	fmt.Fprintf(writer, header)
	for _, t := range sortTeams(teams) {
		name := t.name
		mp := t.matches
		w := t.wins
		d := t.draws
		l := t.losses
		p := t.points
		fmt.Fprintf(writer, "%-30s | %2d | %2d | %2d | %2d | %2d\n", name, mp, w, d, l, p)
	}

	return nil
}
